test/dependency-manager/interpolation-validation.test.ts

```diff
@@ -0,0 +1,196 @@
+        const component_config = `
+        name: hello-world
+        services:
+          api:
+            build:
+              args:
+                \${{ if architect.environment == 'local' }}:
+                  ENV: local
+        `
+
+        const component_spec = buildSpecFromYml(component_config)
+        registerInterpolation(component_spec, context)
+      });
+
+      it('can use conditional around build block if local', async () => {
+        const component_config = `
+        name: hello-world
+        services:
+          api:
+            \${{ if architect.environment == 'local' }}:
+              build:
+                args:
+                  ENV: local
+        `
+
+        const component_spec = buildSpecFromYml(component_config)
+        registerInterpolation(component_spec, context)
+      });
+
+      it('can use conditional around service block with build block if local', async () => {
+        const component_config = `
+        name: hello-world
+        services:
+          \${{ if architect.environment == 'local' }}:
+            api:
+              build:
+                args:
+                  ENV: local
+        `
+
+        const component_spec = buildSpecFromYml(component_config)
+        registerInterpolation(component_spec, context)
+      });
+    });
+
+    it('can use tag conditional in build block', async () => {
+      const component_config = `
+        name: hello-world
+        services:
+          api:
+            build:
+              args:
+                \${{ if architect.build.tag == 'latest' }}:
+                  ENV: prod
+        `
+
+      const component_spec = buildSpecFromYml(component_config)
+      registerInterpolation(component_spec, context)
+    });
+
+    it('can use tag in build block', async () => {
+      const component_config = `
+        name: hello-world
+        services:
+          api:
+            build:
+              args:
+                TAG: \${{ architect.build.tag }}
+        `
+
+      const component_spec = buildSpecFromYml(component_config)
+      registerInterpolation(component_spec, context)
+    });
+
+    it('can use secret outside build block', async () => {
+      const component_config = `
+        name: hello-world
+        secrets:
+          test:
+            required: true
+        services:
+          api:
+            environment:
+              TEST: \${{ secrets.test }}
+        `
+
+      const component_spec = buildSpecFromYml(component_config)
+      registerInterpolation(component_spec, context)
+    });
+
+    it('can still use conditional without build block', async () => {
+      const component_config = `
+        name: hello-world
+        services:
+          api:
+            \${{ if architect.environment == 'local' }}:
+              environment:
+                TEST: test
+        `
+
+      const component_spec = buildSpecFromYml(component_config)
+      registerInterpolation(component_spec, context)
+    });
+  });
+});
```

Comment:
I think these test cases make sense, though there's one common pattern that we use ourselves that I don't see here and will be a breaking change to our own stack's config. For example, you can't do something like 

```yml
a-service:
    ...
    build:
      ...
      args:
        NODE_ENV: ${{ services.a-service.environment.NODE_ENV }}
    environment:
      NODE_ENV: production
    ${{ if secrets.environment == 'local' }}:
      environment:
        NODE_ENV: development
```

The error is

```
ValidationErrors
component: cloud
file: architect.yml:153:23
  150 |     build:
  151 |       ...
  152 |       args:
› 153 |         NODE_ENV: ${{ services.a-service.environment.NODE_ENV }}
      |                       ﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋﹋ Cannot use ${{ services.a-service.environment.NODE_ENV }} inside of a build block. Use ${{ if architect.build.tag == 'latest' }}: for build time conditionals.
```

I'd have expected this interpolation to work since it doesn't rely on a secret. Is that a bug or intentional?

Reply:
Good call - let me think about this